src/utils/components/withMatchMedia.jsx

```diff
@@ -0,0 +1,143 @@
+ * @returns {String} - state property name (for example, isAtMediumUp)
+ */
+export const getStateNameByBreakpoint = breakpoint => {
+	const capitalizedBp = breakpoint[0].toUpperCase() + breakpoint.slice(1);
+	return `isAt${capitalizedBp}Up`;
+};
+
+/**
+ * @param {Array} mediaQueries - list of matchMedia-created MediaQueryList objects
+ * @param {Array} breakpoints - array of breakpoint names that were passed to HOC
+ */
+export const getUpdatedMediaState = (mediaQueries, breakpoints) => {
+	const updatedMedia = {};
+	mediaQueries.forEach((mq, i) => {
+		updatedMedia[getStateNameByBreakpoint(breakpoints[i])] = mq.matches;
+	});
+	return updatedMedia;
+};
+
+/**
+ * @param {Array} breakpoints
+ * @throws {Error}
+ * @return {undefined}
+ */
+export const validateBreakpoints = (breakpoints) => {
+	const validBreakpoints = Object.keys(MEDIA_QUERIES);
+
+	if (!breakpoints) {
+		throw new Error('withMatchMedia: breakpoints array required');
+	}
+
+	breakpoints.forEach(bp => {
+		if (!validBreakpoints.includes(bp)) {
+			throw new Error(`withMatchMedia: ${bp} is not a valid breakpoint name`);
+		}
+	});
+};
+
+
+/**
+ * Provides viewport-aware props to wrapped component.
+ *
+ * @param {React.element} InnerComponent - the component to wrap
+ * @param {Array} breakpoints - array of breakpoint names to watch
+ */
+export const withMatchMedia = (
+	InnerComponent,
+	breakpoints
+) => class extends React.Component {
+	/**
+	 * @constructor
+	 * @param {Object} - React element props
+	 */
+	constructor(props) {
+		super(props);
+		validateBreakpoints(breakpoints);
+
+		this.handleMediaChange = this.handleMediaChange.bind(this);
+
+		this.state = {
+			media: {}
+		};
+
+		breakpoints.forEach(bp => {
+			this.state.media[getStateNameByBreakpoint(bp)] = false;
+		});
+	}
+
+	/**
+	 * handles media change for _all_ breakpoints configured
+	 * updates state.media with matched media
+	 * @returns {undefined}
+	 */
+	handleMediaChange() {
+		const updated = getUpdatedMediaState(this.mediaQueries, breakpoints);
+		this.setState({
+			media: {...this.state.media, ...updated}
+		});
+	}
+
+	/**
+	 * react lifecycle method
+	 * sets up media queries and listeners
+	 * @returns {undefined}
+	 */
+	componentDidMount() {
+		if (typeof window.matchMedia != undefined) {
+
+			// map breakpoints to MediaQueryList objects...
+			this.mediaQueries = breakpoints
+				.map(bp => window.matchMedia(MEDIA_QUERIES[bp]));
+
+			// and add a listener for each
+			this.mediaQueries.forEach(mq => {
+				mq.addListener(this.handleMediaChange);
+			});
+
+			// fire media handler immediately on mount to populate `this.state.media`
+			this.handleMediaChange();
+		}
+	}
+
+	/**
+	 * react lifecycle method
+	 * clean up media query listeners on unmount
+	 * @returns {undefined}
+	 */
+	componentWillUnmount() {
+		this.mediaQueries.forEach((mq, i) => {
+			mq.removeListener(this.handleMediaChange);
+			this.mediaQueries.splice(i, 1);
```

Comment:
I'm removing created `MediaQueryList` objects from this array here for the sake of testing, but have had some difficulty actually testing lifecycle methods of `withMatchMedia` itself